pkg/controllers/user/certsexpiration/certsexpiration.go

```diff
@@ -0,0 +1,169 @@
+package certsexpiration
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/rancher/kontainer-engine/cluster"
+	"github.com/rancher/rancher/pkg/controllers/management/clusterprovisioner"
+	rkecluster "github.com/rancher/rke/cluster"
+	"github.com/rancher/rke/pki"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/cert"
+)
+
+type Controller struct {
+	ClusterLister v3.ClusterLister
+	ClusterClient v3.ClusterInterface
+	ClusterStore  cluster.PersistentStore
+	UserContxt    *config.UserContext
+}
+
+func Register(ctx context.Context, userContext *config.UserContext) {
+	c := &Controller{
+		ClusterLister: userContext.Management.Management.Clusters("").Controller().Lister(),
+		ClusterClient: userContext.Management.Management.Clusters(""),
+		ClusterStore:  clusterprovisioner.NewPersistentStore(userContext.Management.Core.Namespaces(""), userContext.Management.Core),
+		UserContxt:    userContext,
+	}
+
+	userContext.Management.Management.Clusters("").AddHandler(ctx, "certificate-expiration", c.sync)
+}
+
+func (c Controller) sync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if key == "" || cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	if cluster.Spec.RancherKubernetesEngineConfig == nil {
+		return cluster, nil
+	}
+	certsExpInfo := map[string]v3.CertExpiration{}
+
+	cluster, err := c.ClusterLister.Get("", key)
+	if err != nil {
+		return cluster, err
+	}
+	certBundle, err := c.getClusterCertificateBundle(cluster.Name)
+	if err != nil {
+		return cluster, err
+	}
+	for certName, certObj := range certBundle {
+		info, err := getCertExpiration(certObj.CertificatePEM)
+		if err != nil {
+			logrus.Debugf("failed to get expiration date for certificate [%s] for cluster [%s]:%v", certName, key, err)
+			continue
+		}
+		certsExpInfo[certName] = info
+	}
+	if !reflect.DeepEqual(cluster.Status.CertificatesExpiration, certsExpInfo) {
+		cluster.Status.CertificatesExpiration = certsExpInfo
+		return c.ClusterClient.Update(cluster.DeepCopy())
+	}
+	return cluster, nil
+
+}
+
+func (c Controller) getClusterCertificateBundle(clusterName string) (map[string]pki.CertificatePKI, error) {
+	// cluster has a state file ?
+	currentState, err := getRKECurrentStateFromStore(c.ClusterStore, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	if currentState != nil {
+		cleanCertificateBundle(currentState.CertificatesBundle)
+		return currentState.CertificatesBundle, nil
+	}
+
+	// No state file, let's try get the certs from the user cluster
+	// clusterContext, err := c.ClusterManager.UserContext(clusterName)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	certs, err := getCertsFromUserCluster(c.UserContxt)
```

Comment:
since we're already running as user controller, we don't need userContext, 
we can just have `secretsLister` in the Controller struct itself